lib/core/src/Cardano/Pool/DB/Sqlite.hs

```diff
@@ -141,15 +136,13 @@ newDBLayer
        -- ^ Database file location, or Nothing for in-memory database
     -> IO (SqliteContext, DBLayer IO)
 newDBLayer trace fp = do
-    let start = startSqliteBackend (ManualMigration mempty) migrateAll trace fp
-    (nMigrations, ctx) <- handlingPersistError trace fp start
-    let dbLayer = mkDBLayer ctx
-    when (nMigrations > Quantity 0) $ do
-        traceWith trace MsgForcedRollback
-        dbLayer & \DBLayer{..} -> atomically $ rollbackTo slotMinBound
-    pure (ctx, dbLayer)
-  where
-    mkDBLayer SqliteContext{runQuery} = DBLayer
+    let io = startSqliteBackend
+            (ManualMigration mempty)
+            (AutoMigrationOnCreateOrRestart migrateAll)
+            trace
+            fp
+    ctx@SqliteContext{runQuery} <- handlingPersistError trace fp io
+    return (ctx, DBLayer
```

Comment:
Hmmm. This would however re-introduce a bug in the stake pool database. For the same reason as in the wallet database (foreign references and cascade deletion), upon automatic migrations of `pool_registration` table, we would loose all data in the `pool_owner` table, creating an inconsistency. So, I am afraid that we also need an `AutoMigrationOnCreate` here too. Which seemingly makes the `AutoMigrationOnCreate` the only actual used branch for the `AutoMigration` data-type :) ... making the type itself somewhat redundant.

Reply:
@KtorZ wrote:

> Hmmm. This would however re-introduce a bug in the stake pool database. For the same reason as in the wallet database (foreign references and cascade deletion), upon automatic migrations of `pool_registration` table, we would loose all data in the `pool_owner` table, creating an inconsistency. So, I am afraid that we also need an `AutoMigrationOnCreate` here too. 

Hmm, I'm not convinced by this argument.

My reason for introducing the `AutoMigration` type is so that we can:

- **Fix** the current behaviour for wallet databases (so that we perform automatic migrations **only** when creating new databases); but
- **Preserve** the current behaviour for pool databases, which is to **always** run migrations, even when attempting to load old databases.

Indeed, with this PR applied, when we attempt to automatically migrate an old pool database, we see the following failure (identical to `master`):
```hs
[cardano-wallet.pools-db:Error:20] [2020-02-11 04:43:25.31 UTC] Failed to migrate the database: : NOT NULL constraint failed: pool_owner_backup.pool_owner_index
```
It's the failure of this automatic migration that triggers our desired **_database reset_**:
```hs
[cardano-wallet.pools-db:Notice:20] [2020-02-11 04:43:25.31 UTC] Non backward compatible database found. Removing old database and re-creating it from scratch. Ignore the previous error.
[cardano-wallet.pools-db:Notice:20] [2020-02-11 04:43:25.32 UTC] 5 migrations were applied to the database.
[cardano-wallet.pools-engine:Info:27] [2020-02-11 04:43:25.33 UTC] Monitoring stake pools. Currently at genesis
[cardano-wallet.pools-engine:Info:27] [2020-02-11 04:43:25.38 UTC] Applying blocks [1242320.6 ... 1259707.4]
[cardano-wallet.pools-engine:Info:27] [2020-02-11 04:43:25.38 UTC] Writing stake-distribution for epoch 1259707
```
This works because we have the following code to handle the automatic migration failure:
```
handlingPersistError trace fp action = action >>= \case
    Right ctx -> pure ctx
    Left _ -> do
        traceWith trace MsgDatabaseReset
        maybe (pure ()) removeFile fp
        action >>= either throwIO pure
```

However, if we change the pool DB to use the `AutoMigrationOnCreate` strategy, then loading a pre-existing old database will no longer cause an automatic migration failure, and we'll no longer trigger the `handlingPersistError` handler, and the DB will no longer be reset, resulting in the following error:
```hs
[cardano-wallet.pools-db:Notice:23] [2020-02-11 04:46:20.75 UTC] No database migrations were necessary.
[cardano-wallet.pools-engine:Error:28] [2020-02-11 04:46:20.76 UTC] Worker has exited unexpectedly: SQLite3 returned ErrorError while attempting to perform prepare "SELECT \"pool_id\", \"slot\", \"margin\", \"cost\" FROM \"pool_registration\" WHERE (\"pool_id\"=?) LIMIT 1": no such table: pool_registration
```

So it seems to me that in the case of pool databases, preserving the current behaviour is actually what we want.

Reply:
@jonathanknowles I am not arguing against the type so-to-speak, the problem isn't the type, the problem is the logic being applied. 
I agree, in case of migration errors, the pool database will reset as expected. But the issue I am talking about (which causes the 0-balance issue) will occurs on successful migrations!

Remember how persistent migrate the SQLite tables:

- copy the content of the table to a temporary table
- drop the table 
- recreate the new modified table
- copy back the data

During the second step, dropping the table will cause some other tables to loose data because of the foreign references and cascading delete. So even upon a successful migration, we may loose data!

Reply:
> But the issue I am talking about (which causes the 0-balance issue) will occurs on successful migrations!

Fair enough. In that case, we can adopt the approach that you suggest, of running automatic migrations _**only when we create a new database**_, for both pools and wallets.

Though for the pool DB, we still need to do a bit of extra work, since disabling automatic migrations for pre-existing DBs causes the following error on startup if we start with an old DB (for example, from last year):
```hs
[cardano-wallet.pools-engine:Error:28] [2020-02-11 04:46:20.76 UTC] Worker has exited unexpectedly: SQLite3 returned ErrorError while attempting to perform prepare "SELECT \"pool_id\", \"slot\", \"margin\", \"cost\" FROM \"pool_registration\" WHERE (\"pool_id\"=?) LIMIT 1": no such table: pool_registration
```

I'll add a patch to fix this.

Reply:
Also, make sure to execute the automatic db migration tests! 

```
rm -rf state-migration-test*
nix-build nix/migration-tests.nix -o migration-tests
./migration-tests/runall.sh
```